tests/test_cache_layout.py

```diff
@@ -119,6 +120,92 @@ def test_file_downloaded_in_cache_with_revision(self):
+
+            expected_reference = "main"
+
+            # Only reference should be `main`.
+            self.assertListEqual(refs, [expected_reference])
+
+            with open(os.path.join(expected_path, "refs", expected_reference)) as f:
+                snapshot_name = f.readline().strip()
+
+            # The `main` reference should point to the only snapshot we have downloaded
+            self.assertListEqual(no_exist_snapshots, [snapshot_name])
+
+            no_exist_path = os.path.join(expected_path, ".no_exist", snapshot_name)
+            no_exist_content = os.listdir(no_exist_path)
+
+            # Only a single file in the no_exist snapshot
+            self.assertEqual(len(no_exist_content), 1)
+
+            # The no_exist content should be our file
+            self.assertEqual(no_exist_content[0], filename)
+
+            with open(os.path.join(no_exist_path, filename)) as f:
+                content = f.read().strip()
+
+            # The contents of the file should be empty.
+            self.assertEqual(content, "")
+
+    def test_no_exist_file_is_cached_with_revision(self):
+        with tempfile.TemporaryDirectory() as cache:
+            hf_hub_download(
+                MODEL_IDENTIFIER, "file_0.txt", cache_dir=cache, revision="file-2"
+            )
+
+            filename = "this_does_not_exist.txt"
+            with self.assertRaises(EntryNotFoundError):
+                # The file does not exist, so we get an exception.
+                hf_hub_download(
+                    MODEL_IDENTIFIER, filename, cache_dir=cache, revision="file-2"
+                )
+
+            expected_directory_name = f'models--{MODEL_IDENTIFIER.replace("/", "--")}'
+            expected_path = os.path.join(cache, expected_directory_name)
+
+            refs = os.listdir(os.path.join(expected_path, "refs"))
+            no_exist_snapshots = os.listdir(os.path.join(expected_path, ".no_exist"))
+
+            expected_reference = "file-2"
+
+            # Only reference should be `file-2`.
+            self.assertListEqual(refs, [expected_reference])
+
+            with open(os.path.join(expected_path, "refs", expected_reference)) as f:
+                snapshot_name = f.readline().strip()
+
+            # The `main` reference should point to the only snapshot we have downloaded
+            self.assertListEqual(no_exist_snapshots, [snapshot_name])
+
+            no_exist_path = os.path.join(expected_path, ".no_exist", snapshot_name)
+            no_exist_content = os.listdir(no_exist_path)
+
+            # Only a single file in the no_exist snapshot
+            self.assertEqual(len(no_exist_content), 1)
+
+            # The no_exist content should be our file
+            self.assertEqual(no_exist_content[0], filename)
+
+            with open(os.path.join(no_exist_path, filename)) as f:
+                content = f.read().strip()
+
+            # The contents of the file should be empty.
+            self.assertEqual(content, "")
```

Comment:
For readability and since this test is literally the same as `test_no_exist_file_is_cached` but with a different `expected_reference` and `revision`, I would factorize it and do parametrized tests.

Either with `with self.subTest():` or `pytest.mark.parametrize()` (if possible in unittest).
First test would have `revision=None` `expected_reference="main"`, second one with `revision="file-2"` `expected_reference="file-2"`.

Reply:
Note that I reproduced what was done for the other cache layout tests in this file:
`test_file_downloaded_in_cache` and `test_file_downloaded_in_cache_with_revision`, so maybe those two should be grouped as well.

Reply:
Thanks for letting me know 👍 
I'll try to avoid as much as possible the code duplication when it's easily doable but let keep mostly unchanged the existing work.